### ISSUE TITLE
ModuleResolution Node16/NodeNext fails without extension starting on 5.5.0-dev.20240413

### ISSUE THREAD

**Issue by Noriller** [microsoft/TypeScript#58214]
### 🔎 Search Terms

moduleResolution node16 nodeNext import extension

### 🕗 Version & Regression Information

- This changed between versions 5.5.0-dev.20240413 and 5.5.0-dev.20240415 (last one checked)

### 💻 Code

```ts
type Something1 = typeof import('@some/package/path/to/file');
type Something2 = typeof import('@some/package/path/to/file.js');
```

### 🙁 Actual behavior

Something1 resolves to `any`
Something2 resolves to `Something`

### 🙂 Expected behavior

Something1 resolves to `Something`
Something2 resolves to `Something`

### Additional information about the issue

This happens on the latest nightly versions and on moduleResolution Node16/NodeNext.

From what I could get, it started on 5.5.0-dev.20240413.

**Comment by RyanCavanaugh** [comment]
We need a complete repro here; the package.json of `@some/package`, the importing package, and the tsconfig for the importing file all matter

**Comment by Noriller** [comment]
https://stackblitz.com/~/github.com/Noriller/ts-issue-58214

I hope this one helps.

In the `test.d.ts` with the current `typescript@next` and same `jsconfig` (its a legacy project, using mostly jsdocs, but with the problem in a `d.ts` file)

If changing ts version with:

` npm i -D typescript@5.5.0-dev.20240412`

then both work

**Comment by jakebailey** [comment]
This bisects to #57896. @andrewbranch

(I'm not an expert, but it seems strange that you'd need an extension when the d.ts file is effectively CJS, importing another CJS file, but importing via a JS file is already funky.)

**Comment by andrewbranch** [comment]
Guessing I accidentally made ImportTypes look like an ESM-mode source location, when they were previously kept CJS-mode for backward compatibility reasons. Will take a look. Thanks!

**Comment by andrewbranch** [comment]
Your jsconfig.json is invalid:

> Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'

`module` is defaulting to es2020 because of `target`, which is causing everything to resolve under ESM rules (since everything will emit as ESM). I’d argue this is probably _more_ logically consistent than what we used to do, where the module resolution implementation made a lot more assumptions about module format and emit.

I guess we could consider making `--moduleResolution node16` imply `--module node16` instead of issuing an error, but I think we did the error on purpose so that people would _notice_ we were making a breaking change when we started requiring these options be used together in https://github.com/microsoft/TypeScript/pull/54567.
